### PR TITLE
Outbound pg factor

### PR DESCRIPTION
**Pull request by calebschoepp**

### CODE REVIEW

File: crates/factor-outbound-pg/Cargo.toml
Comment (rylev) [inline crates/factor-outbound-pg/Cargo.toml:13]: I'm not sure we need variables at all (see my other comment), but if we do, and they're only be used in tests, can we move this to `dev-dependencies`?

File: crates/factor-outbound-pg/Cargo.toml
Comment (rylev) [inline crates/factor-outbound-pg/Cargo.toml:19]:
```suggestion
tokio-postgres = "0.7.7"
```

File: crates/factor-outbound-pg/src/host.rs
Comment (rylev) [inline crates/factor-outbound-pg/src/host.rs:55]: Not sure if this is copied from the current implementation, but it does seem like `is_address_allowed` should be returning an `anyhow::Result<bool>` so that this unwrap isn't necessary.

File: crates/factor-outbound-pg/src/host.rs
Comment (rylev) [inline crates/factor-outbound-pg/src/host.rs:32]: This isn't blocking merging this PR since this is how things work in the current implementation, but it would be really nice if the host implementation here didn't assume a `tokio_postgres::Client`. 

Instead we could have the connection pool keep track of `Box<dyn PgClient>` where `PgClient` in this case is some trait that abstracts over the underlying postgres calls. 

When running this library as part of the wider Spin CLI, the `PgClient` impl would be backed by a `tokio_postgres::Client` but in test it would be backed by some dummy implementation.

This is what we're doing in the `llm` implementation so that tests don't need to run an actual LLM.  

Thoughts?

File: crates/factor-outbound-pg/tests/factor_test.rs
Comment (rylev) [inline crates/factor-outbound-pg/tests/factor_test.rs:13]: Why do we include this here? Does `OutboundNetworkingFactor` unconditionally depend on `VariablesFactor`? I would have assumed that it would only depend on it if a variable was used but perhaps not?

File: crates/factor-outbound-pg/tests/factor_test.rs
Comment (rylev) [inline crates/factor-outbound-pg/tests/factor_test.rs:50]: Nit: such a todo isn't really meaningful in my opinion and should just be removed.

File: crates/factor-outbound-pg/tests/factor_test.rs
Comment (rylev) [inline crates/factor-outbound-pg/tests/factor_test.rs:45]: If you plan on leaving this TODO past this PR please provide more information on what needs to be "fixed" here.

File: crates/factors/tests/smoke.rs
Comment (rylev) [inline crates/factors/tests/smoke.rs:24]: @lann I think we should decide here and now if adding new factors to the smoke test is necessary. If the point of the smoke test is to make sure `#[derive(Factors)]` works, I think we've already accomplished that, and this change shouldn't be necessary.

Comment (lann) [reply crates/factors/tests/smoke.rs:24]: Agreed - I don't think we need to keep updating this test.

File: crates/factor-outbound-pg/tests/factor_test.rs
Comment (lann) [reply crates/factor-outbound-pg/tests/factor_test.rs:13]: Yeah, the dependency is unconditional. Making it optional would require some refactoring in `AllowedHostsConfig` to detect whether `allowed_outbound_hosts` needs variables, and more generally might make factor dependencies more difficult to reason about.

File: crates/factor-outbound-pg/src/host.rs
Comment (lann) [reply crates/factor-outbound-pg/src/host.rs:32]: :+1: It could also be a defaulted generic param on the whole factor (i.e. `OutboundPgFactor<Client = tokio_postgres::Client>`.

File: crates/factor-outbound-pg/tests/factor_test.rs
Comment (calebschoepp) [reply crates/factor-outbound-pg/tests/factor_test.rs:50]: Oh yeah sorry, I'm not done with this PR so this is a TODO for myself to finish up this PR. I marked it draft, but I guess it wasn't super clear b/c I requested reviews.

Agreed that in the abstract this isn't a helpful PR.

File: crates/factor-outbound-pg/tests/factor_test.rs
Comment (calebschoepp) [reply crates/factor-outbound-pg/tests/factor_test.rs:45]: Same as above.

File: crates/factor-outbound-pg/src/host.rs
Comment (calebschoepp) [reply crates/factor-outbound-pg/src/host.rs:32]: Filed an issue to do this work in a followup PR #2649

**Comment by calebschoepp** [comment]
@lann @rylev this is ready for another round of reviews.

**Review by lann — APPROVED** [review]

**Review by rylev — APPROVED** [review]

File: crates/factor-outbound-pg/Cargo.toml
Comment (calebschoepp) [reply crates/factor-outbound-pg/Cargo.toml:16]: We use it for the spawn

**Pull request merged**
